### ISSUE TITLE
QN ITensor constructor API

### ISSUE THREAD

**Issue by mtfishman** [ITensor/ITensors.jl#349]
In the process of writing documentation for the QN ITensor constructors, it made me wonder about the interface. There are the following sets of constructors:
```julia
ITensor(::QN, ::QNIndex...)
diagITensor(::QN, ::QNIndex...)
delta(::QN, ::QNIndex...)
```
where the fluxes are specified. These are all well defined. However, there are also the following possible constructors:
```julia
ITensor(::QNIndex...)
diagITensor(::QNIndex...)
delta(::QNIndex...)
```
where the flux is not specified, and the question I have is how they should be defined.

In the case of `diagITensor` and `delta`, it seems natural to define them as:
```julia
diagITensor(::QNIndex...) = diagITensor(QN(), ::QNIndex...)
delta(::QNIndex...) = delta(QN(), ::QNIndex...)
```
since it is common to want the diagonal blocks specified when you have matching indices like `i', dag(i)`. This definition also helps make code between ITensors and QN ITensors more generic. In contrast, right now `ITensor(::QNIndex...) != ITensor(QN(), ::QNIndex...)`. Instead, it makes an ITensor with `BlockSparse` storage and no blocks.

I believe the current behavior of the `diagITensor` and `delta` constructors are pretty natural and useful. I was wondering if it made sense to change the definition for the `ITensor` constructor to `ITensor(::QNIndex...) = ITensor(QN(), ::QNIndex...)`. I have often thought it might be useful to have that, and that convention could also be used to define `randomITensor(::QNIndex...) = randomITensor(QN(), ::QNIndex...)`, which I have thought could be convenient.

Another reason for the change is that in light of the proposal in #340, the current use cases of `ITensor(::QNIndex...)` would be superseded by the constructor `zeroITensor(::QNIndex...)`. I think `zeroITensor(::QNIndex...)` could act the same way and be more generic, and it would be strange to have two constructors that do (essentially) the same thing. @emstoudenmire, thoughts on this would be appreciated!

**Comment by emstoudenmire** [comment]
It's a good question. I'm very much for the functions `diagITensor` and `delta` to assume a flux of zero as a default, because it certainly makes a lot of sense there. It also makes sense for `ITensor` and thereby `randomITensor`, because zero-flux ITensors are the most common type, in a sense, and it's a universal kind of default which can always succeed. It would also foster generic code (this may be the strongest argument of all), where we don't have to do a check like `hasqns` just to know if we should put `QN()` into some ITensor constructor.

But one reservation we should think about is whether it could lead to surprising behavior for users. The main example that came to mind would be like if they wanted to make a random product state for, say, S=1/2 degrees of freedom. Then calling `randomITensor` on those product tensors wouldn't work, and it could be surprising. Whereas it would work for S=1 and would make a Sz=0 product state. Maybe that case isn't so important - I was just trying to think about different cases.

**Comment by mtfishman** [comment]
That was my main reservation, that people might use `ITensor(::QNIndex...)` and `randomITensor(::QNIndex...)`, which works many times until it doesn't (and people aren't even realizing they are making ITensors with flux `QN()`). However, I think it would be clear what is going on, since if people start setting the wrong elements, we give an error that the flux is wrong and people will immediately see that it was making flux `QN()` ITensors by default. We can even point them to the `zeroITensor(::QNIndex...)` constructor in the `setindex!` error message which should be encouraged as the generic way to initialize ITensors.

I think it will help a lot with generic code. I like the trick of changing indices from `Index(2)` to `Index(QN() => 2)` as a way to "turn on" arrow checking. In that case, having the constructors `ITensor` and `randomITensor` default to flux `QN()` _should_ make all of the following code generic for ITensors and QN ITensors.

**Comment by emstoudenmire** [comment]
In my opinion, having it help a lot with generic code makes it worth doing, so I think we should enact this change. The other worries are not so bad, because we can deal with them with good documentation and error messages. 

The documentation can clearly state that for the QN case, randomITensor and ITensor default to flux-zero, if not specified, and that an error will arise if the indices don't admit a flux-zero case. And the error messages can similarly say "no way to make a flux-zero ITensor; please specify a flux" or similar.

**Comment by mtfishman** [comment]
I was actually considering returning an ITensor with no blocks if there are no zero flux blocks, instead of an error (I think the code now would just naturally handle that case). I'm not sure which way is better. It seems like someone could easily catch that no blocks were made if they wanted to avoid that case.

**Comment by emstoudenmire** [comment]
My instinct here is to kind of split the difference between too strict and too loose. So we are agreed that always requring a QN is perhaps too strict, in part because it can make generic code much more complicated. 

But I'd argue that always returning an ITensor, even with no blocks, is too loose. The reason is that if it causes an error down the line, it will be hard for someone, even us, to backtrack and figure out why the error happened. Some number of lines later an ITensor will not have any blocks, and then one has to know that it's because randomITensor (say) didn't put any in. But then it still might not be totally obvious to a new person why randomITensor didn't put in any.

Another way to look at it is: isn't the main reason to call randomITensor at all is to get a non-zero tensor? Just where it's not clear or obvious which value other than zero is good, but definitely not zero.

Finally, we could always remove this error message later (the error message saying randomITensor with default flux would result in no blocks).

**Comment by mtfishman** [comment]
I agree that calling `randomITensor` and getting no blocks would be strange, but there may be legitimate cases where one creates an `ITensor` and ends up with no blocks (though if you mean to do that, then the current canonical way would be using `emptyITensor`).

Sounds good to keep it as an error for now and consider allowing it if we come across a reasonable use case.

**Comment by emstoudenmire** [comment]
Ok great, yes let’s keep it as an error and can always revisit later.

**Comment by mtfishman** [comment]
This was implemented in #348.
